### PR TITLE
Add mediator and remove AddTextNotification from OrderManager

### PR DESCRIPTION
**Pull request by teinarss**

### CODE REVIEW

**Comment by pchote** [comment]
There's a lot of overlap between `Mediator` and our `TypeDictionary`. It would be good to think about whether we can either use `TypeDictionary` as-is, otherwise what kind of changes would be needed for that to work in this case.

**Comment by teinarss** [comment]
Updated with using `TypeDictionary`, 

Todo:
* Clear NotificationsCache when the game has ended?

**Comment by teinarss** [comment]
Updated.

**Comment by Mailaender** [comment]
Needs a rebase
> No description provided.

and some hints what is actually achieved here.

**Comment by PunkPun** [comment]
I could potentially use this to fix https://github.com/OpenRA/OpenRA/pull/20100#discussion_r916687431

**Comment by teinarss** [comment]
rebased.

File: OpenRA.Game/TextNotificationsManager.cs [hunk 1]
```diff
@@ -82,5 +91,10 @@ static bool IsPoolEnabled(TextNotificationPool pool)
 				(pool == TextNotificationPool.Transients && filters.HasFlag(TextNotificationPoolFilters.Transients)) ||
 				(pool == TextNotificationPool.Feedback && filters.HasFlag(TextNotificationPoolFilters.Feedback));
 		}
+
+		public static void Dispose()
```
Comment (PunkPun) [inline OpenRA.Game/TextNotificationsManager.cs]: `TextNotificationsManager` isn't `IDisposable` so the method should not be called `Dispose`

**Review by PunkPun — CHANGES_REQUESTED** [review]
This works ok in lobby but crashes instantly ingame when you write a message

<details closed>
<summary>Crash</summary>
Exception of type `System.InvalidOperationException`: TypeDictionary contains multiple instances of type `OpenRA.Widgets.INotificationHandler`1[OpenRA.TextNotification]`
   at OpenRA.Primitives.TypeDictionary.Get(Type t, Boolean throwsIfMissing) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Primitives/TypeDictionary.cs:line 64
   at OpenRA.Primitives.TypeDictionary.Get[T]() in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Primitives/TypeDictionary.cs:line 51
   at OpenRA.Network.UnitOrders.ProcessOrder(OrderManager orderManager, World world, Int32 clientId, Order order) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Network/UnitOrders.cs:line 30
   at OpenRA.Network.OrderManager.ReceiveImmediateOrders(Int32 clientId, OrderPacket orders) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Network/OrderManager.cs:line 166
   at OpenRA.Network.NetworkConnection.OpenRA.Network.IConnection.Receive(OrderManager orderManager) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Network/Connection.cs:line 287
   at OpenRA.Sync.<>c__DisplayClass13_0.<RunUnsynced>b__0() in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Sync.cs:line 167
   at OpenRA.Sync.RunUnsynced[T](Boolean checkSyncHash, World world, Func`1 fn) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Sync.cs:line 205
   at OpenRA.Sync.RunUnsynced(World world, Action fn) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Sync.cs:line 168
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Game.cs:line 624
   at OpenRA.Game.LogicTick() in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Game.cs:line 639
   at OpenRA.Game.Loop() in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Game.cs:line 804
   at OpenRA.Game.Run() in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Game.cs:line 857
   at OpenRA.Game.InitializeAndRun(String[] args) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Game/Game.cs:line 296
   at OpenRA.Launcher.Program.Main(String[] args) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Launcher/Program.cs:line 32
</details>

ps. you didn't rebase on latest bleed but this was good enough for me

**Comment by teinarss** [comment]
Updated.

File: OpenRA.Game/Widgets/Widget.cs [hunk 2]
```diff
@@ -530,7 +547,10 @@ public virtual void Removed()
 
 			if (LogicObjects != null)
 				foreach (var l in LogicObjects)
+				{
+					Ui.Unsubscribe(l);
 					l.Dispose();
+				}
```
Comment (abcdefg30) [inline OpenRA.Game/Widgets/Widget.cs]: Missing braces on the outer `if`.

File: OpenRA.Test/OpenRA.Game/MediatorTest.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
+﻿using NUnit.Framework;
```
Comment (abcdefg30) [inline OpenRA.Test/OpenRA.Game/MediatorTest.cs]: Missing header.

File: OpenRA.Game/Game.cs [hunk 4]
```diff
@@ -83,6 +83,8 @@ public static string TimestampedFilename(bool includemilliseconds = false, strin
 
 		static void JoinInner(OrderManager om)
 		{
+			TextNotificationsManager.Clear();
```
Comment (PunkPun) [inline OpenRA.Game/Game.cs:87]: Perhaps add a comment here, it feels rather out of place. Perhaps something like this or more indepth

> Refresh TextNotificationsManager before the game starts

fwiw I'm planning to insert MutedPlayers.Clear() into this method

File: OpenRA.Game/TextNotificationsManager.cs [hunk 5]
```diff
@@ -20,6 +21,9 @@ public static class TextNotificationsManager
 		static readonly string SystemMessageLabel;
 
 		public static long ChatDisabledUntil { get; internal set; }
+		static readonly List<TextNotification> NotificationsCache = new List<TextNotification>();
+
+		public static IReadOnlyList<TextNotification> Notifications => NotificationsCache;
```
Comment (PunkPun) [inline OpenRA.Game/TextNotificationsManager.cs:25]: Just a small nit, IMO there shouldn't be a space between `NotificationsCache` and `Notifications`

**Review by PunkPun — COMMENTED** [review]
LGTM, just a few suggestions

**Comment by teinarss** [comment]
Updated

**Review by PunkPun — DISMISSED** [review]

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs [hunk 6]
```diff
@@ -260,7 +256,7 @@ public void CloseChat()
 			Ui.ResetTooltips();
 		}
 
-		public void AddNotificationWrapper(TextNotification notification)
+		public void Handle(TextNotification notification)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs]: Please implement these interfaces explicitly if possible:
```suggestion
		void INotificationHandler<TextNotification>.Handle(TextNotification notification)
```
(here and the two others below.)
That makes the code more readable by indicating where the `Handle` method belongs to.

**Review by abcdefg30 — DISMISSED** [review]
Lgtm otherwise.

**Comment by teinarss** [comment]
Fixes

**Review by PunkPun — APPROVED** [review]

**Comment by PunkPun** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/8d968c292f84280c72ff57e98a2d748c9c94566f...ec33cf4e85e04f1a97c9ac024a0257d08dc104fa

**Pull request merged**